lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,1039 @@
+		return ErrConnectTimeout
+
+	case <-s.doneReceiving:
+		return ErrNotConnected
+	}
+}
+
+// getNextRequestID returns the next request ID that should be used when sending
+// a request to the remote signer.
+func (s *SignCoordinator) getNextRequestID() uint64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	nextRequestID := s.nextRequestID
+
+	s.nextRequestID++
+
+	return nextRequestID
+}
+
+// createResponseChannel creates a response channel for the given request ID and
+// inserts it into the responses map. The function returns a cleanup function
+// which removes the channel from the responses map, and the caller must ensure
+// that this cleanup function is executed once the thread that's waiting for
+// the response is done.
+func (s *SignCoordinator) createResponseChannel(requestID uint64) func() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	respChan := make(chan *walletrpc.SignCoordinatorResponse, 1)
+
+	// Insert the response channel into the map.
+	s.responses[requestID] = respChan
+
+	// Create a cleanup function that will delete the response channel.
+	return func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		select {
+		case <-respChan:
+			// If we have timed out, there could be a very unlikely
+			// scenario where we did receive a response before we
+			// managed to grab the lock in the cleanup func.
+			// In that case, we'll just ignore the response.
+			// We should still clean up the response channel though.
+		default:
+		}
+
+		delete(s.responses, requestID)
+	}
+}
+
+// getResponse waits for a response with the given request ID and returns the
+// response if it is received. If the corresponding response from the remote
+// signer is a SignerError, the error message is returned. If the response is
+// not received within the given timeout, an error is returned.
+//
+// Note: Before calling this function, the caller must have created a response
+// channel for the request ID.
+func (s *SignCoordinator) getResponse(requestID uint64,
+	timeout time.Duration) (*walletrpc.SignCoordinatorResponse, error) {
+
+	s.mu.Lock()
+
+	// Verify that we have a response channel for the request ID.
+	if _, ok := s.responses[requestID]; !ok {
+		// It should be impossible to reach this case, as we create the
+		// response channel before sending the request.
+		s.mu.Unlock()
+
+		return nil, fmt.Errorf("no response channel found for "+
+			"request ID %d", requestID)
+	}
+
+	// Grab the response channel for the request ID.
+	respChan := s.responses[requestID]
+
+	s.mu.Unlock()
+
+	// Wait for the response to arrive.
+	select {
+	case resp, ok := <-respChan:
+		if !ok {
+			// If the response channel was closed, we return an
+			// error as the receiving thread must have timed out
+			// before we managed to grab the response.
+			return nil, ErrRequestTimeout
+		}
+
+		// If the response is an error, we return the error message.
+		//nolint:lll
+		if errorResp, ok := resp.GetSignResponseType().(*walletrpc.SignCoordinatorResponse_SignerError); ok {
```

Comment:
nit: could create a type alias just above this line to somewhat limit the length.